[git-webkit] Fix issue assignee failure

youssefsoli opened the pull request:
#### 601fa2d162a1e694c3e0f2d395f944c0cb0be452
<pre>
[git-webkit] Fix issue assignee failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=242324">https://bugs.webkit.org/show_bug.cgi?id=242324</a>

Reviewed by Jonathan Bedard.

The GitHub API incorrectly isn&apos;t able to assign assignees to an issue
through the issue PATCH endpoint for certain user accounts, however, when
making the request to the issues/{id}/assignees endpoint, the assignment
is successful.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:

Canonical link: <a href="https://commits.webkit.org/252159@main">https://commits.webkit.org/252159@main</a>
</pre>

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
```diff
@@ -404,14 +408,26 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
 
         return self.add_comment(issue, why) if why else issue
 
+    def add_assignees(self, issue, assignees):
```
This looks to be `set_assignees`, not `add_assignees`, right?

youssefsoli replied:
Making a POST request to that endpoint doesn't override the assignees array, it just adds any assignees that haven't already been added. Not sure what the verb for that would be.

JonWBedard replied:
`add` is right, in that case, just a bit surprised that GitHub's API makes this `add` rather than `set`. Documentation reference for others curious: [https://docs.github.com/en/rest/issues/assignees#add-assignees-to-an-issue](https://docs.github.com/en/rest/issues/assignees#add-assignees-to-an-issue)

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
```diff
@@ -404,14 +408,26 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
+        )
+        if not response:
+            issue._assignee = None
+            return False
```
Should return an empty list, I think, to be consistent.

youssefsoli replied:
The `all()` function returns True if everything in the generator is True, otherwise it returns False. I am considering switching it to return the list of assignees if that would make more sense here.

JonWBedard commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
```diff
@@ -334,7 +334,11 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
-            update_dict['assignees'] = [issue._assignee.username]
+            assignees = [issue._assignee.username]
+            if not self.add_assignees(issue, assignees):
+                sys.stderr.write('Could not add assignee(s) to issue')
```
Seems like `add_assignees` should own the error here.

youssefsoli replied:
Makes sense.

JonWBedard reviewed (commented):
webkitpy test failures are real here. [Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/github.py](https://github.com/WebKit/WebKit/blob/main/Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/github.py) has the mock web server that we're using in unit testing, I think you just need to mock the assignees endpoint and everything should just work.

youssefsoli commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
```diff
@@ -334,7 +334,9 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
-            update_dict['assignees'] = [issue._assignee.username]
+            assignees = [issue._assignee.username]
+            if self.add_assignees(issue, assignees) != assignees:
+                return None
```
Not sure if the whole tool should fail if assignment isn't successful, previously the error was never caught.

JonWBedard replied:
I had this same thought.

Ultimately, it's a pretty rare situation that would require someone making a single call that bot set the assignee and did something else. While that's possible (and relevant for bugzilla and radar, where such operations _are_ done in a single request), it's a rare enough situation that I'm not inclined to advocate increasing code complexity to deal with it.

youssefsoli committed:
[git-webkit] Fix issue assignee failure
https://bugs.webkit.org/show_bug.cgi?id=242324

Reviewed by Jonathan Bedard.

The GitHub API incorrectly isn't able to assign assignees to an issue
through the issue PATCH endpoint for certain user accounts, however, when
making the request to the issues/{id}/assignees endpoint, the assignment
is successful.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:

Canonical link: https://commits.webkit.org/252159@main
